router/src/server/http/write_test_helpers.rs

```diff
@@ -0,0 +1,262 @@
+    (
+        $name:ident,
+        encoding = $encoding:tt,
+        uri = $uri:expr,
+        body = $body:expr,
+        dml_info_handler = $dml_info_handler:ident,
+        dml_write_handler = $dml_write_handler:expr,
+        dml_delete_handler = $dml_delete_handler:expr,
+        want_result = $want_result:pat,
+        want_dml_calls = $($want_dml_calls:tt )+
+    ) => {
+        paste::paste! {
+            mod [<test_mod_ test_http_handler_ $name _ $encoding>] {
+                use super::*;
+                use $crate::{
+                    test_http_handler,
+                    dml_handlers::{
+                        mock::{MockDmlHandler},
+                    },
+                    namespace_resolver::{mock::MockNamespaceResolver},
+                    server::http::{
+                        self,
+                        HttpDelegate,
+                        WriteInfoExtractor,
+                        write_test_helpers::assert_metric_hit,
+                    },
+                };
+                use assert_matches::assert_matches;
+                #[allow(unused_imports)]
+                use flate2::{write::GzEncoder, Compression};
+                #[allow(unused_imports)]
+                use hyper::{header::{CONTENT_ENCODING, HeaderValue}, Body, Request, StatusCode};
+                #[allow(unused_imports)]
+                use std::{io::Write, sync::Arc};
+
+                #[tokio::test]
+                async fn [<test_http_handler_ $name _ $encoding>]() {
+                    let body = $body;
+
+                    // Optionally generate a fragment of code to encode the body
+                    let body = test_http_handler!(encoding=$encoding, body);
+
+                    #[allow(unused_mut)]
+                    let mut request = Request::builder()
+                        .uri($uri)
+                        .method("POST")
+                        .body(Body::from(body))
+                        .unwrap();
+
+                    // Optionally modify request to account for the desired
+                    // encoding
+                    test_http_handler!(encoding_header=$encoding, request);
+
+                    // add all the mappings used in the tests.
+                    let mock_namespace_resolver = MockNamespaceResolver::default()
+                        .with_mapping("bananas_test", NAMESPACE_ID)
+                        .with_mapping("test", NAMESPACE_ID)
+                        .with_mapping("database", NAMESPACE_ID)
+                        .with_mapping("database/myrp", NAMESPACE_ID);
+                    let dml_handler = Arc::new(MockDmlHandler::default()
+                        .with_write_return($dml_write_handler)
+                        .with_delete_return($dml_delete_handler)
+                    );
+                    let metrics = Arc::new(metric::Registry::default());
+                    let dml_info_extractor: &'static dyn WriteInfoExtractor = &http::[<$dml_info_handler>];
```

Comment:
Yeah, this works to get the correct dml_info_handler struct. 🤯

Reply:
This is correct - it's tricky because you're initialising a type and it might not be in scope where the macro is called - scopes are evaluated when the macro is _called_ and not when it is defined (unlike everything else!)

So this:

```rust
use crate::http::CST;

test_http_handler!(..., CST);
```

is different from:

```rust
test_http_handler!(..., CST);
```

I think you probably want to change the macro inputs to:

```
dml_info_handler = $dml_info_handler:expr, // An expression, that returns a value
```

And change this to use the value:
```suggestion
                    let dml_info_extractor: &'static dyn WriteInfoExtractor = &$dml_info_handler;
```

And then call it like this:
```rust
use crate::http::CST;

test_http_handler!(..., CST::default());
```

Which will a) have far less scope for...scope problems (🥁) and b) allow constructing these CST/MT structs with impl specific parameters if/whem they're added (such as CST `autocreate_database`).

Ultimately you've got a real raw deal with these tests as your first PR - feel free to comment them out and we can tackle them together at the end :+1:

Reply:
Definitely a cleaner approach. Updated. TY.

I noticed the scope weirdness, but it felt hard to keep track of scope from the chain of callers. Tried instead to move the scope imports into the macros themselves (refer to macro `test_http_handler!`), but that did make some weirdness. That macro expansion rule now creates another mod, with the scope imports, and then the test function. Is this the pattern acceptable for the scope needed for the final `async fn [<test_http_handler_ $name _ $encoding>]()` test?